src/ocean/util/container/btree/BTree.d

```diff
@@ -0,0 +1,695 @@
+        x.elements = the elements themselves
+        x.is_leaf = indicator if the node is the leaf.
+        x.child_nodes = number_of_elements+1 pointers to its children
+
+    - Elements in the node split the ranges of the elements stored
+      in each subtree.
+
+    - All leaves have the same depth
+
+    - Every node other than root must have at least (degree-1) elements.
+
+    - Every internal node other than the root has at least degree children.
+
+    - Every node may contain at most 2*degree - 1 keys. Therefore, an internal
+      node may have at most 2*degree children.
+
+    Other than insert and delete operations, inorder traversal method is
+    added, which traverses the tree in inorder order (left subtree is visited
+    first, then the element, then the right subtree). This provides sequential
+    access to the tree. There's a recursive and range-based implementation
+    of the inorder traversal.
+
+    CPU complexity of the operations (n is number of the elements, t is degree,
+        h is the tree height):
+
+        - Searching: O(th) = O(t * log_t(n))
+        - Inserting: O(th) = O(t * log_t(n))
+        - Deleting:  O(th) = O(t * log_t(n))
+
+    Complexity of the memory access (to read/write the node from/in memory):
+
+        - Searching O(h) = O(log_t(n))
+        - Inserting O(h) = O(log_t(n))
+        - Deleting: O(h) = O(log_t(n))
+
+    References:
+        - https://en.wikipedia.org/wiki/B-tree
+        - Thomas H. Cormen et al. "Introduction to Algorithms, 3rd edition"
+
+    Params:
+        TreeElementType = type of the element to store.
+        tree_degree = degree of a tree (refer to the documentation)
+        Allocator = allocation technique to use
+
+******************************************************************************/
+
+struct BTree(TreeElementType, int tree_degree, alias Allocator = MallocAllocator,
+        alias TrackingPolicy = TrackAllocationsPolicy)
+{
+    /***********************************************************************
+
+        Recursive inorder iteration. Note that the iteration is undefined
+        in case the tree is changed during the iteration.
+
+        Params:
+            dg = opApply's delegate
+
+        Returns:
+            return value of dg
+
+    ***********************************************************************/
+
+    public int opApply (int delegate (ref ElementType) dg)
+    {
+        return inorder(*this, dg);
+    }
+
+    /**************************************************************************
+
+        Inserts the element in the tree.
+
+        Params:
+            el = element to insert.
+
+        Returns:
+            pointer to a copy of the element in the tree, if the element
+            didn't already exist, null otherwise.
+
+    ***************************************************************************/
+
+    public Const!(ElementType)* insertElement (ElementType el)
```

Comment:
I thought about this one, and I think it needs to stay as it is. In case of value-types, this would create a copy of the value type, which we'll put into the tree. In case of the reference type, we will do the same - this will provide the copy of the reference and it would be put into the tree as well. My feeling about this being `const` is that it's too harsh - the object itself doesn't need to be const, _except_ when it comes to ordering - ordering needs to stay the same. Templating `BTree` on `Const!(Object)` would provide the guarantee to the user, if they wish. What do you think?